KeyError: 'status'  if bz['status'] in OPEN_STATUSES:

KeyError: 'status'  if bz['status'] in OPEN_STATUSES: 
```
2019-11-15 13:07:52 - robottelo.helpers - DEBUG - Generated file bz_cache.json with BZ collect data

tests/foreman/cli/test_repository.py::RepositoryTestCase::test_negative_synchronize_docker_repo_with_mix_valid_invalid_tags FAILED                                    [100%]

================================================================================= FAILURES ==================================================================================
___________________________________________ RepositoryTestCase.test_negative_synchronize_docker_repo_with_mix_valid_invalid_tags ____________________________________________

self = <tests.foreman.cli.test_repository.RepositoryTestCase testMethod=test_negative_synchronize_docker_repo_with_mix_valid_invalid_tags>

    @tier2
    def test_negative_synchronize_docker_repo_with_mix_valid_invalid_tags(self):
        """Set tags whitelist to contain both valid and invalid (non-existing)
        tags. Check if only whitelisted tags are synchronized
    
        :id: 75668da8-cc94-4d39-ade1-d3ef91edc812
    
        :expectedresults: Only whitelisted tag is synchronized
        """
        tags = ['latest', gen_string('alpha')]
        repo = self._make_repository({
            'content-type': 'docker',
            'docker-upstream-name': 'alpine',
            'url': DOCKER_REGISTRY_HUB,
            'docker-tags-whitelist': ",".join(tags),
        })
        Repository.synchronize({'id': repo['id']})
>       repo = self._validated_image_tags_count(repo=repo)

tests/foreman/cli/test_repository.py:1064: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
tests/foreman/cli/test_repository.py:132: in _validated_image_tags_count
    if is_open('BZ:1664631'):
robottelo/helpers.py:779: in is_open
    return handlers[handler_code.strip()](issue.strip(), data)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

issue = 'BZ:1664631', data = None

    def is_open_bz(issue, data=None):
        """Check if specific BZ is open consulting a cached `data` dict or
        calling Bugzilla REST API.
    
        Arguments:
            issue {str} -- The BZ reference e.g: BZ:123456
            data {dict} -- Issue data indexed by <handler>:<number> or None
        """
    
        bz = try_from_cache(issue, data)
        if bz.get("is_open") is not None:  # bug has been already processed
            return bz["is_open"]
    
        bz = follow_duplicates(bz)
    
        # BZ is explicitly in OPEN status
>       if bz['status'] in OPEN_STATUSES:
E       KeyError: 'status'

robottelo/issue_handlers/bugzilla.py:41: KeyError
```

--- a/robottelo/helpers.py
+++ b/robottelo/helpers.py
@@ -753,6 +753,22 @@ def host_provisioning_check(ip_addr):
                 result.stdout))
 
 
+def slugify_component(string, keep_hyphens=True):
+    """Make component name a slug
+
+    Arguments:
+        string {str} -- Component name e.g: ActivationKeys
+        keep_hyphens {bool} -- Keep hyphens or replace with underscores
+
+    Returns:
+        str -- component slug e.g: activationkeys
+    """
+    string = string.replace(" and ", "&")
+    if not keep_hyphens:
+        string = string.replace('-', '_')
+    return re.sub("[^-_a-zA-Z0-9]", "", string.lower())
+
+
 # --- Issue based Pytest markers ---
 
 def is_open(issue, data=None):
@@ -790,6 +806,16 @@ def _should_deselect(issue, data=None):
     if str(issue).startswith(supported_handlers):
         handler_code, _, _ = str(issue).partition(":")
         return handlers[handler_code.strip()](issue.strip(), data)
+
+
+def _add_workaround(
+    data, matches, usage, validation=(lambda *a, **k: True), **kwargs
+):
+    """Adds entry for workaround usage."""
+    for match in matches:
+        issue = f"{match[0]}:{match[1]}"
+        if validation(data, issue, usage, **kwargs):
+            data[issue.strip()]['used_in'].append({'usage': usage, **kwargs})
 
 
 def generate_issue_collection(items, config):  # pragma: no cover
@@ -862,23 +888,46 @@ def generate_issue_collection(items, config):  # pragma: no cover
         re.IGNORECASE
     )
 
+    IMPORTANCE = re.compile(
+        # To match :CaseImportance: Critical
+        r"\s*:CaseImportance:\s*(?P<importance>\S*)",
+        re.IGNORECASE
+    )
+
+    BZ = re.compile(
+        # To match :BZ: 123456, 456789
+        r"\s*:BZ:\s*(?P<bz>.*\S*)",
+        re.IGNORECASE
+    )
+
+    test_modules = set()
+
     # --- Build the issue marked usage collection ---
     for item in items:
-        filepath, lineno, testcase = item.location
-        # Take the module level :CaseComponent:
-        module_component = None
-        with open(item.module.__file__, 'r') as modulefile:
-            text = modulefile.read()
-            matches = COMPONENT.findall(text)
-            if matches:
-                module_component = matches[0]
-        # Then take from source of function if exists
-        source = inspect.getsource(item.function)
-        if ':CaseComponent:' in source:
-            matches = COMPONENT.findall(source)
-            if matches:
-                module_component = matches[0]
+        component = None
+        bzs = None
+        importance = None
 
+        # register test module as processed
+        test_modules.add(item.module)
+
+        # Find matches from docstrings top-down from: module, class, function.
+        mod_cls_fun = (item.module, getattr(item, 'cls', None), item.function)
+        for docstring in map(inspect.getdoc, mod_cls_fun):
+            if not docstring:
+                continue
+            component_matches = COMPONENT.findall(docstring)
+            if component_matches:
+                component = component_matches[-1]
+            bz_matches = BZ.findall(docstring)
+            if bz_matches:
+                bzs = bz_matches[-1]
+            importance_matches = IMPORTANCE.findall(docstring)
+            if importance_matches:
+                importance = importance_matches[-1]
+
+        filepath, lineno, testcase = item.location
+        component_mark = slugify_component(component, False)
         for marker in item.iter_markers():
             if marker.name in valid_markers:
                 issue = marker.kwargs.get('reason') or marker.args[0]
@@ -888,30 +937,88 @@ def generate_issue_collection(items, config):  # pragma: no cover
                         'filepath': filepath,
                         'lineno': lineno,
                         'testcase': testcase,
-                        'component': module_component,
+                        'component': component,
+                        'importance': importance,
+                        'component_mark': component_mark,
                         'usage': marker.name
                     }
                 )
 
                 # Store issue key to lookup in the deselection process
                 deselect_data[item.location] = issue_key
+                # Add issue as a marker to enable filtering e.g: "-m BZ_123456"
+                item.add_marker(
+                    getattr(pytest.mark, issue_key.replace(':', '_'))
+                )
 
         # Then take the workarounds using `is_open` helper.
+        source = inspect.getsource(item.function)
         if 'is_open(' in source:
-            def add_workaround(matches, condition):
-                for match in matches:
-                    issue = f"{match[0]}:{match[1]}"
-                    collected_data[issue.strip()]['used_in'].append(
-                        {
-                            'filepath': filepath,
-                            'lineno': lineno,
-                            'testcase': testcase,
-                            'component': module_component,
-                            'usage': condition
-                        }
-                    )
-            add_workaround(IS_OPEN.findall(source), 'is_open')
-            add_workaround(NOT_IS_OPEN.findall(source), 'not is_open')
+            kwargs = {
+                'filepath': filepath,
+                'lineno': lineno,
+                'testcase': testcase,
+                'component': component,
+                'importance': importance,
+                'component_mark': component_mark
+            }
+            _add_workaround(
+                collected_data,
+                IS_OPEN.findall(source),
+                'is_open',
+                **kwargs
+            )
+            _add_workaround(
+                collected_data,
+                NOT_IS_OPEN.findall(source),
+                'not is_open',
+                **kwargs
+            )
+
+        # Add component as a marker to anable filtering e.g: "-m contentviews"
+        item.add_marker(getattr(pytest.mark, component_mark))
+
+        # Add BZs from tokens as a marker to enable filter e.g: "-m BZ_123456"
+        if bzs:
+            for bz in bzs.split(','):
+                item.add_marker(getattr(pytest.mark, f'BZ_{bz.strip()}'))
+
+        # Add importance as a token
+        if importance:
+            item.add_marker(getattr(pytest.mark, importance.lower().strip()))
+
+    # Take uses of `is_open` from outside of test cases e.g: SetUp methods
+    for test_module in test_modules:
+        module_source = inspect.getsource(test_module)
+        component_matches = COMPONENT.findall(module_source)
+        module_component = None
+        if component_matches:
+            module_component = component_matches[0]
+        if 'is_open(' in module_source:
+            kwargs = {
+                'filepath': test_module.__file__,
+                'lineno': 1,
+                'testcase': test_module.__name__,
+                'component': module_component,
+            }
+
+            def validation(data, issue, usage, **kwargs):
+                return issue not in data
+
+            _add_workaround(
+                collected_data,
+                IS_OPEN.findall(module_source),
+                'is_open',
+                validation=validation,
+                **kwargs
+            )
+            _add_workaround(
+                collected_data,
+                NOT_IS_OPEN.findall(module_source),
+                'not is_open',
+                validation=validation,
+                **kwargs
+            )
 
     # --- Collect BUGZILLA data ---
     bugzilla.collect_data_bz(collected_data, cached_data)

--- a/robottelo/issue_handlers/bugzilla.py
+++ b/robottelo/issue_handlers/bugzilla.py
@@ -241,19 +241,7 @@ def get_data_bz(bz_numbers, cached_data=None):  # pragma: no cover
             "Provide api_key or a bz_cache.json."
         )
         # Provide default data for collected BZs
-        return [
-            {
-                "id": number,
-                "is_open": True,  # All marked is skipped
-                "is_deselected": False,  # nothing is deselected
-                "status": "",
-                "resolution": "",
-                "clone_ids": [],
-                "cf_clone_of": "",
-                "error": "missing bugzilla api_key"
-            }
-            for number in bz_numbers
-        ]
+        return [get_default_bz(number) for number in bz_numbers]
 
     # No cached data so Call Bugzilla API
     LOGGER.debug(f"Calling Bugzilla API for {set(bz_numbers)}")
@@ -299,6 +287,21 @@ def get_single_bz(number, cached_data=None):  # pragma: no cover
         try:
             bz_data = cached_data[f"BZ:{number}"]['data']
         except (KeyError, TypeError):
-            bz_data = get_data_bz([str(number)], cached_data)[0]
+            bz_data = get_data_bz([str(number)], cached_data)
+            bz_data = bz_data and bz_data[0]
         CACHED_RESPONSES['get_single'][number] = bz_data
-    return bz_data or {}
+    return bz_data or get_default_bz(number)
+
+
+def get_default_bz(number):  # pragma: no cover
+    """This is the default BZ data when it is not possible to reach BZ api"""
+    return {
+        "id": number,
+        "is_open": True,  # All marked is skipped
+        "is_deselected": False,  # nothing is deselected
+        "status": "",
+        "resolution": "",
+        "clone_ids": [],
+        "cf_clone_of": "",
+        "error": "missing bugzilla api_key"
+    }

--- a/tests/foreman/conftest.py
+++ b/tests/foreman/conftest.py
@@ -20,7 +20,7 @@ def log(message, level="DEBUG"):
     so we need to emulate the logger by stdouting the output
     """
     now = datetime.datetime.utcnow()
-    full_message = "{date} - conftest - {level} - {message}\n".format(
+    full_message = "{date} - conftest - {level} - {message}".format(
         date=now.strftime("%Y-%m-%d %H:%M:%S"),
         level=level,
         message=message
@@ -150,13 +150,7 @@ def pytest_collection_modifyitems(items, config):
         if skip_if_open:
             # marker must have `BZ:123456` as argument.
             issue = skip_if_open.kwargs.get('reason') or skip_if_open.args[0]
-            if is_open(issue):
-                log(
-                    f'Skipping {item.function.__name__} '
-                    f'in {item.function.__module__} '
-                    f'due to {issue}'
-                )
-                item.add_marker(pytest.mark.skip(reason=issue))
+            item.add_marker(pytest.mark.skipif(is_open(issue), reason=issue))
 
     config.hook.pytest_deselected(items=deselected_items)
     items[:] = [item for item in items if item not in deselected_items]
@@ -184,6 +178,11 @@ def pytest_configure(config):
     ]
     for marker in markers:
         config.addinivalue_line("markers", marker)
+
+    # ignore warnings about dynamically added markers e.g: component markers
+    config.addinivalue_line(
+        'filterwarnings', 'ignore::pytest.PytestUnknownMarkWarning'
+    )
 
 
 def pytest_addoption(parser):

--- a/tests/robottelo/test_helpers.py
+++ b/tests/robottelo/test_helpers.py
@@ -1,27 +1,29 @@
 """Tests for module ``robottelo.helpers``."""
 # (Too many public methods) pylint: disable=R0904
 import os
-import six
+
 import unittest2
-from robottelo.helpers import (
-    HostInfoError,
-    escape_search,
-    get_host_info,
-    get_server_version,
-    Storage,
-    is_open,
-    _should_deselect
-)
+
+from collections import defaultdict
+
 from robottelo.constants import (
     CLOSED_STATUSES,
     OPEN_STATUSES,
     WONTFIX_RESOLUTIONS
 )
+from robottelo.helpers import (
+    HostInfoError,
+    Storage,
+    _add_workaround,
+    _should_deselect,
+    escape_search,
+    get_host_info,
+    get_server_version,
+    is_open,
+    slugify_component
+)
 
-if six.PY2:
-    import mock
-else:
-    from unittest import mock
+from unittest import mock
 
 
 class GetServerVersionTestCase(unittest2.TestCase):
@@ -89,10 +91,7 @@ class GetHostInfoTestCase(unittest2.TestCase):
         ssh.command = mock.MagicMock(return_value=FakeSSHResult([''], 0))
         with self.assertRaises(HostInfoError) as context:
             get_host_info()
-        if six.PY2:
-            message = context.exception.message
-        else:
-            message = str(context.exception)
+        message = str(context.exception)
         self.assertEqual(message, 'Not able to parse release string ""')
 
 
@@ -106,7 +105,7 @@ class FakeSSHResult(object):
 class EscapeSearchTestCase(unittest2.TestCase):
     def test_return_type(self):
         """Tests if escape search returns a unicode string"""
-        self.assertIsInstance(escape_search('search term'), six.text_type)
+        self.assertIsInstance(escape_search('search term'), str)
 
     def test_escapes_double_quotes(self):
         """Tests if escape search escapes double quotes"""
@@ -474,3 +473,39 @@ class BugzillaIssueHandlerTestCase(unittest2.TestCase):
         for issue in ("BZ123456", "XX:123456", "KK:89456", "123456", 999999):
             with self.subTest(issue=issue):
                 self.assertIsNone(_should_deselect(issue))
+
+
+def test_slugify_component():
+    """Assert slugify_component returns proper values"""
+    assert slugify_component('ContentViews') == 'contentviews'
+    assert slugify_component('File-Management') == 'file-management'
+    assert slugify_component('File-Management', False) == 'file_management'
+    assert slugify_component('File&Management') == 'filemanagement'
+    assert slugify_component('File and Management') == 'filemanagement'
+
+
+def test_add_workaround():
+    """Assert helper function adds corrent items to given data"""
+    data = defaultdict(lambda: {"data": {}, "used_in": []})
+    matches = [('BZ', '123456'), ('BZ', '789456')]
+
+    _add_workaround(
+        data,
+        matches,
+        'test',
+        foo='bar'
+    )
+
+    _add_workaround(
+        data,
+        matches,
+        'test',
+        validation=lambda *a, **k: False,  # Should not be added
+        zaz='traz'
+    )
+
+    for match in matches:
+        issue = f"{match[0]}:{match[1]}"
+        used_in = data[issue.strip()]['used_in']
+        assert {'usage': 'test', 'foo': 'bar'} in used_in
+        assert {'usage': 'test', 'zaz': 'traz'} not in used_in

